File: backend/app/schemas/reward.py
"""Reward response schemas."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

RewardType = Literal["integer", "float", "boolean"]
Difficulty = Literal["easy", "medium", "hard", "flexible"]


# ── Totals (existing) ──────────────────────────────────────────────


class MemberRewardTotals(BaseModel):
    """Per-member totals across all approved submissions."""

    member_id: uuid.UUID
    name: str
    totals: dict[str, float]


class RewardsResponse(BaseModel):
    """Response for GET /rewards."""

    members: list[MemberRewardTotals]


# ── Reward category CRUD ───────────────────────────────────────────


class RewardCategoryCreateRequest(BaseModel):
    """Body for POST /rewards/categories."""

    name: str = Field(min_length=1, max_length=255)
    type: RewardType


class RewardCategoryPatchRequest(BaseModel):
    """Body for PATCH /rewards/categories/{id}."""

    name: str | None = Field(default=None, min_length=1, max_length=255)
    type: RewardType | None = None


class RewardCategoryRead(BaseModel):
    """Public view of a reward category."""

    model_config = ConfigDict(from_attributes=True)

    reward_category_id: uuid.UUID
    household_id: uuid.UUID
    name: str
    type: str
    created_at: datetime


# ── Difficulty→reward mapping CRUD ─────────────────────────────────


class DifficultyMappingUpsert(BaseModel):
    """Single item in the bulk PUT /rewards/mappings body."""

    difficulty: Difficulty
    reward_category_id: uuid.UUID
    value: float


class DifficultyMappingRead(BaseModel):
    """Public view of a difficulty→reward mapping."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    household_id: uuid.UUID
    difficulty: str
    reward_category_id: uuid.UUID
    value: float
    created_at: datetime

File: flutter_app/lib/core/models/reward.dart
class MemberRewardTotals {
  MemberRewardTotals({
    required this.memberId,
    required this.name,
    required this.totals,
  });

  final String memberId;
  final String name;
  final Map<String, double> totals;

  factory MemberRewardTotals.fromJson(Map<String, dynamic> json) =>
      MemberRewardTotals(
        memberId: json['member_id'] as String,
        name: json['name'] as String,
        totals: (json['totals'] as Map<String, dynamic>)
            .map((k, v) => MapEntry(k, (v as num).toDouble())),
      );
}

class RewardCategory {
  RewardCategory({
    required this.rewardCategoryId,
    required this.householdId,
    required this.name,
    required this.type,
    required this.createdAt,
  });

  final String rewardCategoryId;
  final String householdId;
  final String name;
  final String type;
  final DateTime createdAt;

  factory RewardCategory.fromJson(Map<String, dynamic> json) => RewardCategory(
        rewardCategoryId: json['reward_category_id'] as String,
        householdId: json['household_id'] as String,
        name: json['name'] as String,
        type: json['type'] as String,
        createdAt: DateTime.parse(json['created_at'] as String),
      );
}

class DifficultyMapping {
  DifficultyMapping({
    required this.id,
    required this.householdId,
    required this.difficulty,
    required this.rewardCategoryId,
    required this.value,
    required this.createdAt,
  });

  final String id;
  final String householdId;
  final String difficulty;
  final String rewardCategoryId;
  final double value;
  final DateTime createdAt;

  factory DifficultyMapping.fromJson(Map<String, dynamic> json) =>
      DifficultyMapping(
        id: json['id'] as String,
        householdId: json['household_id'] as String,
        difficulty: json['difficulty'] as String,
        rewardCategoryId: json['reward_category_id'] as String,
        value: (json['value'] as num).toDouble(),
        createdAt: DateTime.parse(json['created_at'] as String),
      );

  Map<String, dynamic> toUpsertJson() => {
        'difficulty': difficulty,
        'reward_category_id': rewardCategoryId,
        'value': value,
      };
}
